### ISSUE TITLE
Feature Request: Support setting the Plejd Network System Clock

### ISSUE THREAD

**Issue by kodarn** [icanos/hassio-plejd#130]
#### Background:
Plejd devices have an internal clock used for dealing with internal automations. This clock is synchronized each timed the user connects to the Plejd-Bluetooth-network using the Plejd App.

![01 BluetoothSystemTimeSync](https://user-images.githubusercontent.com/24460172/103560540-39774980-4eb8-11eb-8ec4-11f493b71db7.png)
_Screenshot of Plejd Bluetooth Network Time synchronization_

#### When power failure occurs:
When a power failure occurs, this clock looses it settings, and any Plejd-device-internal automations are disabled. This state remains until the user reconnects to the Plejd Bluetooth network using the Plejd App, and the clock is synchronized again.

![02 BluetoothSystemTimeSyncDone](https://user-images.githubusercontent.com/24460172/103560573-43994800-4eb8-11eb-8b4b-22c92a4688a5.png)
_Screenshot of popup after a power failure_

#### Steps to repeat:

1. Disable power to all of your Plejd devices
2. Enable power to all of your Plejd devices
3. Connect to your Plejd Bluetooth network using the Plejd App

## Suggested feature
Implement support for sending the `SET-SYSTEM-TIME` message in hassio-plejd to the devices.

#### Minor note:
Plejd have a physical battery backup solution called BAT-01 https://www.plejd.com/sv-se/produkter/BAT-01 to avoid having to reconnect to the Bluetooth network after power failures.

**Comment by SweVictor** [comment]
With the newest code changes it should be possible to pick up the BLE message sent by the Plejd app to update the clock if you set logLevel to verbose or silly.

After that it's the matter of decoding the actual data that is sent and see if we can replicate it. 

Anyone up to getting some logs? Logs should nowadays be timestamped which should give a good hint of what the message _should_ contain!

On a side note I wonder what happens if a user has the Gateway AND we update the clock. Fight?! :)

**Comment by emilohman** [comment]
I'm the creator of the Plejd app for Homey and node-red. I don't use Home assistant so I write here instead of giving you a pull request.
I've ported Klas python code (https://github.com/klali/ha-plejd/commit/66603fcf4a944a0dc3d069f781f8301dc9691b50) to JS that sets the system clock in Plejd.

Plejd will broadcast the time every 5 or 10 minute like this:
010110001b2aeb236001

| broadcast | don't respond | time command | the time |
| --------- | ------------- | ------------ | -------- |
| 01 | 0110 | 001b | 2aeb236001 |

but if Plejd loose it's power the internal clock will be reset to 0 and start ticking from there (if you don't have BAT-01).
When this happens Plejd stops broadcasting the time. To fix this either ask Plejd for it's time or just always set the time when your app starts.

Here is some code to get you started (but needs to be fixed to fit your code):

```
/*
 * Plejd reports time without timezone.
 */
const getUTCTime = (date) => {
  return date.getTime() - (date.getTimezoneOffset() * 60000);
};

/*
 * Ask a device in the mesh for the current time.
 * Eg: 0b0102001b: 0b: id, 0102: read, 001b: time
 */
const timeGet = (id) => {
  const payload = Buffer.from((id).toString(16).padStart(2, '0') + '0102001b', 'hex');

  return encodeDecode(this.cryptokey, this.address, payload);
};

/*
 * Set current time.
 * E.g: 000110001b38df236000: 00: set?, 0110: don't respond, 001b: time command, 38df236000: the time
 */
const timeSet = () => {
  const now = getUTCTime(new Date());

  const buffer = Buffer.alloc(10);
  buffer.write('000110001b', 0, 'hex');
  buffer.writeInt32LE(Math.trunc(now / 1000), 5);
  buffer.write('00', 9, 'hex');

  return encodeDecode(this.cryptokey, this.address, buffer);
};

/*
 * Parse incoming time command. Plejd time is whitout timezone and comes as seconds (not milliseconds).
 * E.g:
 * 0b0103001b37df236000 for response from device 11: 0b: id, 0103: read respone?, 001b: time command, 37df236000: the time
 * 010110001b2aeb236001 for broadcast: 01: broadcast, 0110: don't respond, 001b: time command, 2aeb236001: the time
 */
const timeParse = (data) => {
  const decodedData = encodeDecode(this.cryptokey, this.address, data);

  if (decodedData.toString('hex', 0, 1) === '01' || decodedData.toString('hex', 3, 5) === '001b') {
    const now = getUTCTime(new Date());
    const time = decodedData.readInt32LE(5) * 1000;

    let diff;
    if (now > time) {
      diff = now - time;
    } else {
      diff = time - now;
    }

    return { time, diff }
  }

  return null;
};

/*
 * Get current time from Plejd. If diff is more than a minute set current time
 */
const syncTime = () => {
  // Ask a device in the mesh for current time. Broadcast dosen't work.
  await this.plejdWrite(timeGet(devices[0].plejdId));

  const data = await this.dataCharacteristic.read();

  const time = timeParse(data);

  if (time) {
     // Sync if diff is more than a minute
    if (time.diff > 60000) {
      await this.plejdWrite(timeSet());
    }
  }
};
```

I hope this helps to get it to work in Home assistant. Just ask if got any problems!

**Comment by SweVictor** [comment]
VERY nice! Thanks a lot for that! So... Now we have THREE home automation projects for Plejd BLE, that's interesting!

I have been thinking for a while that it would be really nice to try and document the Plejd BLE and possibly also BLE. The above information you posted is really nice, but trying to figure that out manually is a REAL pain!

Has anyone started anything like this? Any interest?

**Comment by SweVictor** [comment]
Moved documenting discussion to #163

**Comment by SweVictor** [comment]
Code to implement this in branch https://github.com/icanos/hassio-plejd/tree/feature/time

With a gateway you can get very strange fighting to set time, and it seems like Plejd devices recalculate (at least) time-based scene states every time a new time update is received. So - what happened to me is that the outside lights started flashing (click on then off) quite frequently. So this functionality should be used with caution and probably not in conjunction with the gateway.

**Comment by icanos** [comment]
Maybe we could detect a gateway, from the response of the API, and in that case, automatically disable the set time functionality? To avoid these issues I mean.

**Comment by SweVictor** [comment]
Possibly. The latest code adds an opt-in option to enable hourly time sync. I think I prefer that to detection-based 'magic' personally

**Comment by SweVictor** [comment]
Resolved in #165, preparing 0.6.1 release tag
